simulator/func_sim/instr_cache.h

```diff
@@ -0,0 +1,53 @@
+/*
+* This is an open source non-commercial project. Dear PVS-Studio, please check it.
+* PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+*/
+/**
+ * A cache for decoded instructions in functional simulator
+ * @author Denis Los
+*/
+
+
+#ifndef INSTR_CACHE_H
+#define INSTR_CACHE_H
+
+#include <map>
+
+template <typename K, typename V>
+class InstrCache
+{
+    public:
+        using iterator = typename std::map<K, V>::iterator;
+
+        iterator begin() { return cache.begin(); }
+        iterator end() { return cache.end(); }
```

Comment:
Should it be `const`? If yes, can we use automated type deduction?

`auto begin() const { return cache.cbegin(); }`
`auto end() const { return cache.cend(); }`